### PR TITLE
First version

### PR DESCRIPTION
**Pull request by mathi123**
First version of OdooC# repository.
v1:
- Annotate C# classes, referring to Odoo models
- Annotate C# properties, referring to Odoo fields, One2many fields and Many2one fields.
- Read/Write/update classes -> odoo rows get updated.

Todo:
- Lazy load One2many and Many2one properties automagically

### CODE REVIEW

File: xmlrpc/XmlRpcStruct.cs [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+      }
+
+      public bool MoveNext()
+      {
+        _index++;
+        if (_index >= _keys.Count)
+          return false;
+        else
+          return true;
+      }
```
Comment (JayVora-SerpentCS) [inline xmlrpc/XmlRpcStruct.cs:176]: 'else' statement is not needed!

File: OdooRpcWrapper/Exceptions/MissingOdooObjectAttributeException.cs [hunk 2]
```diff
@@ -0,0 +1,18 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using System.Runtime.Serialization;
+
```
Comment (JayVora-SerpentCS) [inline OdooRpcWrapper/Exceptions/MissingOdooObjectAttributeException.cs:7]: We may reduce the line of code by unncecessary imports!

File: xmlrpc/XmlRpcStruct.cs (hunk 1 above)
Comment (pedrobaeza) [inline xmlrpc/XmlRpcStruct.cs:176]: Even more, this can be simplified as `return _index >= _keys.Count`

File: OdooRpcWrapper/OdooAPI_Search.cs [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+            Type type = typeof(T);
+
+            CheckLogin();
+            CheckModel(type);
+
+            string model = OdooObjectAttribute.GetModelName(type);
+
+            return _objectRpc.search(_credentials.DbName, _credentials.UserId,
+                                     _credentials.DbPassword, model, "search", filter);
+        }
```
Comment (JayVora-SerpentCS) [inline OdooRpcWrapper/OdooAPI_Search.cs:18]: We may here think of the additional params like limit,offset, order by!

**Comment by JayVora-SerpentCS** [comment]
Overall,your effort is excellent!
Give some time for review as the lines of code are many!

File: xmlrpc/XmlRpcStruct.cs (hunk 1 above)
Comment (mathi123) [inline xmlrpc/XmlRpcStruct.cs:176]: The xmlrpc project is the raw version of the (famous) .NET xmlrpc library from Cook Computing. I guess this project does not really need review. 
(It was included so Travis could build correctly)

**Comment by bwrsandman** [comment]
Did you finally get Travis working?

You think you can cleanup the commits? Things like Travis 1, 2, ..., 10 could probably be squashed into a single commit. Same thing with the update readme and update docs commits.

The idea is to have each commit be a logical change or addition to the code, the commits fixing earlier commits of the pull request aren't that necessary in the git history.

You can use `git rebase --interactive` to clean up the commits.

File: .travis.yml [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+  apt:
+    packages:
+      - expect-dev  # provides unbuffer utility
+      - python-lxml  # because pip installation is slow
+      - python-simplejson
+      - python-serial
+
+env:
+  global:
+  - VERSION="8.0"
```
Comment (bwrsandman) [inline .travis.yml:15]: Are you planning to support other versions?
If that's the case, you may want to set up a env matrix for the various versions and not use global version.

``` yml
env:
  matrix:
  - VERSION="6.1"
  - VERSION="7.0"
  - VERSION="8.0"
```

**Comment by mathi123** [comment]
yes I did :smile: 
The problem was in the xplrpc library that I used. By including the source, it is now recompiled through mono on travis, and my unit tests can run.

File: .travis.yml [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+  - createdb test
+  - git clone https://github.com/mathi123/maintainer-quality-tools.git ${HOME}/maintainer-quality-tools -b unhardcode
+  - export PATH=${HOME}/maintainer-quality-tools/travis:${PATH}
+  - travis_install_nightly
+  - nuget restore OdooDotNet.sln
+  - nuget install NUnit.Runners -Version 2.6.4 -OutputDirectory testrunner
+
+script:
+  - /home/travis/odoo-8.0/openerp-server -d test -i crm,sale --stop-after-init --workers=0 --log-level=warn
+  - /home/travis/odoo-8.0/openerp-server -d test --log-level=warn &
```
Comment (bwrsandman) [inline .travis.yml]: It's best to factor out the version now instead of later for other versions

``` bash
/home/travis/odoo-${VERSION}/openerp-server
```

**Comment by bwrsandman** [comment]
If you can, add a linter to the tests.

I recommend you come up with a code style for this project and add a CONTRIBUTING.md with the style guide.

It will help out in the long run and avoid future discussion about it.

**Comment by mathi123** [comment]
Cleaner PR [here](https://github.com/OCA/dotnet/pull/2)!

**Pull request closed without merging**